ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java

```diff
@@ -3136,6 +3136,55 @@ public void testBlobDelete503NoRetrySinceNoSuccessReplica() throws Exception {
     }
   }
 
+  /**
+   * Test Deletion first hits Not_FOUND. Then run ReplicateBlob.
+   * If the PutBlob is compacted already, ReplicateBlob may fail with ID_Deleted.
+   * @throws Exception
+   */
+  @Test
+  public void testReplicateBlobOnDeleteButPutBlobCompacted() throws Exception {
+    String serviceID = "delete-service";
+    MockServerLayout layout = new MockServerLayout(mockClusterMap);
+    String localDcName = "DC3";
+    setRouter(getNonBlockingRouterProperties(localDcName), layout, new LoggingNotificationSystem());
+    setOperationParams();
+    String blobId = router.putBlob(putBlobProperties, putUserMetadata, putChannel, new PutBlobOptionsBuilder().build())
+        .get(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+
+    // two local replicas are unavailable. Remote replica doesn't have the Blob.
+    DataNodeId sourceDataNode = null;
```

Comment:
sourceDataNode is not used except for assignment at L3163

Reply:
test passes even if i remove it

Reply:
ReplicateBlob needs a source replica to replicate the Blob from. So at least we should have one replica available.

Removing the code won't have on-demand replication on-going.

Reply:
sourceDataNode is a local variable here and not passed as an arg anywhere. How is replicateBlob using it internally ?

Reply:
also, how do we know replicateBlob actually took place ?

Reply:
oh i see, what u mean. By falling into the if-case u r ensuring not all servers throw an error.

Reply:
so by removing it, the test passed because on-demand repl was not happening. Hmm, this is confusing. The test shd check if on-demand repl actually happened. Is there a way to do that ?

Reply:
Yes